File: client/src-tauri/src/commands/contracts/autonomous.rs
use serde::{Deserialize, Serialize};

use super::runtime::RuntimeRunDiagnosticDto;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousRunStatusDto {
    Starting,
    Running,
    Paused,
    Cancelling,
    Cancelled,
    Stale,
    Failed,
    Stopped,
    Crashed,
    Completed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousRunRecoveryStateDto {
    Healthy,
    RecoveryRequired,
    Terminal,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousUnitKindDto {
    Researcher,
    Planner,
    Executor,
    Verifier,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousUnitStatusDto {
    Pending,
    Active,
    Blocked,
    Paused,
    Completed,
    Cancelled,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousUnitArtifactStatusDto {
    Pending,
    Recorded,
    Rejected,
    Redacted,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousToolCallStateDto {
    Pending,
    Running,
    Succeeded,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousVerificationOutcomeDto {
    Passed,
    Failed,
    Blocked,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousSkillLifecycleStageDto {
    Discovery,
    Install,
    Invoke,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousSkillLifecycleResultDto {
    Succeeded,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AutonomousSkillCacheStatusDto {
    Miss,
    Hit,
    Refreshed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousCommandResultDto {
    pub exit_code: Option<i32>,
    pub timed_out: bool,
    pub summary: String,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum GitToolResultScopeDto {
    Staged,
    Unstaged,
    Worktree,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum WebToolResultContentKindDto {
    Html,
    PlainText,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CommandToolResultSummaryDto {
    pub exit_code: Option<i32>,
    pub timed_out: bool,
    pub stdout_truncated: bool,
    pub stderr_truncated: bool,
    pub stdout_redacted: bool,
    pub stderr_redacted: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct FileToolResultSummaryDto {
    pub path: Option<String>,
    pub scope: Option<String>,
    pub line_count: Option<usize>,
    pub match_count: Option<usize>,
    pub truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct GitToolResultSummaryDto {
    pub scope: Option<GitToolResultScopeDto>,
    pub changed_files: usize,
    pub truncated: bool,
    pub base_revision: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WebToolResultSummaryDto {
    pub target: String,
    pub result_count: Option<usize>,
    pub final_url: Option<String>,
    pub content_kind: Option<WebToolResultContentKindDto>,
    pub content_type: Option<String>,
    pub truncated: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case", tag = "kind")]
pub enum ToolResultSummaryDto {
    Command(CommandToolResultSummaryDto),
    File(FileToolResultSummaryDto),
    Git(GitToolResultSummaryDto),
    Web(WebToolResultSummaryDto),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousToolResultPayloadDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub attempt_id: String,
    pub artifact_id: String,
    pub tool_call_id: String,
    pub tool_name: String,
    pub tool_state: AutonomousToolCallStateDto,
    pub command_result: Option<AutonomousCommandResultDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_summary: Option<ToolResultSummaryDto>,
    pub action_id: Option<String>,
    pub boundary_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousVerificationEvidencePayloadDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub attempt_id: String,
    pub artifact_id: String,
    pub evidence_kind: String,
    pub label: String,
    pub outcome: AutonomousVerificationOutcomeDto,
    pub command_result: Option<AutonomousCommandResultDto>,
    pub action_id: Option<String>,
    pub boundary_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousPolicyDeniedPayloadDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub attempt_id: String,
    pub artifact_id: String,
    pub diagnostic_code: String,
    pub message: String,
    pub tool_name: Option<String>,
    pub action_id: Option<String>,
    pub boundary_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousSkillLifecycleSourceDto {
    pub repo: String,
    pub path: String,
    pub reference: String,
    pub tree_hash: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousSkillLifecycleCacheDto {
    pub key: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<AutonomousSkillCacheStatusDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousSkillLifecycleDiagnosticDto {
    pub code: String,
    pub message: String,
    pub retryable: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousSkillLifecyclePayloadDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub attempt_id: String,
    pub artifact_id: String,
    pub stage: AutonomousSkillLifecycleStageDto,
    pub result: AutonomousSkillLifecycleResultDto,
    pub skill_id: String,
    pub source: AutonomousSkillLifecycleSourceDto,
    pub cache: AutonomousSkillLifecycleCacheDto,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub diagnostic: Option<AutonomousSkillLifecycleDiagnosticDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case", tag = "kind")]
pub enum AutonomousArtifactPayloadDto {
    ToolResult(AutonomousToolResultPayloadDto),
    VerificationEvidence(AutonomousVerificationEvidencePayloadDto),
    PolicyDenied(AutonomousPolicyDeniedPayloadDto),
    SkillLifecycle(AutonomousSkillLifecyclePayloadDto),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousLifecycleReasonDto {
    pub code: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousRunDto {
    pub project_id: String,
    pub run_id: String,
    pub runtime_kind: String,
    pub provider_id: String,
    pub supervisor_kind: String,
    pub status: AutonomousRunStatusDto,
    pub recovery_state: AutonomousRunRecoveryStateDto,
    pub active_unit_id: Option<String>,
    pub active_attempt_id: Option<String>,
    pub duplicate_start_detected: bool,
    pub duplicate_start_run_id: Option<String>,
    pub duplicate_start_reason: Option<String>,
    pub started_at: String,
    pub last_heartbeat_at: Option<String>,
    pub last_checkpoint_at: Option<String>,
    pub paused_at: Option<String>,
    pub cancelled_at: Option<String>,
    pub completed_at: Option<String>,
    pub crashed_at: Option<String>,
    pub stopped_at: Option<String>,
    pub pause_reason: Option<AutonomousLifecycleReasonDto>,
    pub cancel_reason: Option<AutonomousLifecycleReasonDto>,
    pub crash_reason: Option<AutonomousLifecycleReasonDto>,
    pub last_error_code: Option<String>,
    pub last_error: Option<RuntimeRunDiagnosticDto>,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousWorkflowLinkageDto {
    pub workflow_node_id: String,
    pub transition_id: String,
    pub causal_transition_id: Option<String>,
    pub handoff_transition_id: String,
    pub handoff_package_hash: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousUnitDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub sequence: u32,
    pub kind: AutonomousUnitKindDto,
    pub status: AutonomousUnitStatusDto,
    pub summary: String,
    pub boundary_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub workflow_linkage: Option<AutonomousWorkflowLinkageDto>,
    pub started_at: String,
    pub finished_at: Option<String>,
    pub updated_at: String,
    pub last_error_code: Option<String>,
    pub last_error: Option<RuntimeRunDiagnosticDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousUnitAttemptDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub attempt_id: String,
    pub attempt_number: u32,
    pub child_session_id: String,
    pub status: AutonomousUnitStatusDto,
    pub boundary_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub workflow_linkage: Option<AutonomousWorkflowLinkageDto>,
    pub started_at: String,
    pub finished_at: Option<String>,
    pub updated_at: String,
    pub last_error_code: Option<String>,
    pub last_error: Option<RuntimeRunDiagnosticDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousUnitArtifactDto {
    pub project_id: String,
    pub run_id: String,
    pub unit_id: String,
    pub attempt_id: String,
    pub artifact_id: String,
    pub artifact_kind: String,
    pub status: AutonomousUnitArtifactStatusDto,
    pub summary: String,
    pub content_hash: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<AutonomousArtifactPayloadDto>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousUnitHistoryEntryDto {
    pub unit: AutonomousUnitDto,
    pub latest_attempt: Option<AutonomousUnitAttemptDto>,
    #[serde(default)]
    pub artifacts: Vec<AutonomousUnitArtifactDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AutonomousRunStateDto {
    pub run: Option<AutonomousRunDto>,
    pub unit: Option<AutonomousUnitDto>,
    pub attempt: Option<AutonomousUnitAttemptDto>,
    #[serde(default)]
    pub history: Vec<AutonomousUnitHistoryEntryDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct GetAutonomousRunRequestDto {
    pub project_id: String,
}

use super::runtime::RuntimeRunControlInputDto;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StartAutonomousRunRequestDto {
    pub project_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub initial_controls: Option<RuntimeRunControlInputDto>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub initial_prompt: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CancelAutonomousRunRequestDto {
    pub project_id: String,
    pub run_id: String,
}

File: client/src/lib/xero-model/autonomous.ts
import { z } from 'zod'
import {
  runtimeRunControlInputSchema,
  runtimeRunDiagnosticSchema,
  type RuntimeRunDiagnosticDto,
} from './runtime'
import { isoTimestampSchema, nonEmptyOptionalTextSchema, normalizeOptionalText, normalizeText } from './shared'

export const autonomousRunStatusSchema = z.enum([
  'starting',
  'running',
  'paused',
  'cancelling',
  'cancelled',
  'stale',
  'failed',
  'stopped',
  'crashed',
  'completed',
])
export const autonomousRunRecoveryStateSchema = z.enum(['healthy', 'recovery_required', 'terminal', 'failed'])

export const autonomousLifecycleReasonSchema = z
  .object({
    code: z.string().trim().min(1),
    message: z.string().trim().min(1),
  })
  .strict()

export const autonomousRunSchema = z
  .object({
    projectId: z.string().trim().min(1),
    agentSessionId: z.string().trim().min(1),
    runId: z.string().trim().min(1),
    runtimeKind: z.string().trim().min(1),
    providerId: z.string().trim().min(1),
    supervisorKind: z.string().trim().min(1),
    status: autonomousRunStatusSchema,
    recoveryState: autonomousRunRecoveryStateSchema,
    duplicateStartDetected: z.boolean(),
    duplicateStartRunId: nonEmptyOptionalTextSchema,
    duplicateStartReason: nonEmptyOptionalTextSchema,
    startedAt: isoTimestampSchema,
    lastHeartbeatAt: nonEmptyOptionalTextSchema,
    lastCheckpointAt: nonEmptyOptionalTextSchema,
    pausedAt: nonEmptyOptionalTextSchema,
    cancelledAt: nonEmptyOptionalTextSchema,
    completedAt: nonEmptyOptionalTextSchema,
    crashedAt: nonEmptyOptionalTextSchema,
    stoppedAt: nonEmptyOptionalTextSchema,
    pauseReason: autonomousLifecycleReasonSchema.nullable().optional(),
    cancelReason: autonomousLifecycleReasonSchema.nullable().optional(),
    crashReason: autonomousLifecycleReasonSchema.nullable().optional(),
    lastErrorCode: nonEmptyOptionalTextSchema,
    lastError: runtimeRunDiagnosticSchema.nullable().optional(),
    updatedAt: isoTimestampSchema,
  })
  .strict()

export const getAutonomousRunRequestSchema = z
  .object({
    projectId: z.string().trim().min(1),
    agentSessionId: z.string().trim().min(1),
  })
  .strict()

export const startAutonomousRunRequestSchema = z
  .object({
    projectId: z.string().trim().min(1),
    agentSessionId: z.string().trim().min(1),
    initialControls: runtimeRunControlInputSchema.nullable().optional(),
    initialPrompt: z.string().trim().min(1).nullable().optional(),
  })
  .strict()

export const cancelAutonomousRunRequestSchema = z
  .object({
    projectId: z.string().trim().min(1),
    agentSessionId: z.string().trim().min(1),
    runId: z.string().trim().min(1),
  })
  .strict()

export const autonomousRunStateSchema = z
  .object({
    run: autonomousRunSchema.nullable(),
  })
  .strict()

export type AutonomousRunStatusDto = z.infer<typeof autonomousRunStatusSchema>
export type AutonomousRunRecoveryStateDto = z.infer<typeof autonomousRunRecoveryStateSchema>
export type AutonomousLifecycleReasonDto = z.infer<typeof autonomousLifecycleReasonSchema>
export type AutonomousRunDto = z.infer<typeof autonomousRunSchema>
export type GetAutonomousRunRequestDto = z.infer<typeof getAutonomousRunRequestSchema>
export type StartAutonomousRunRequestDto = z.infer<typeof startAutonomousRunRequestSchema>
export type CancelAutonomousRunRequestDto = z.infer<typeof cancelAutonomousRunRequestSchema>
export type AutonomousRunStateDto = z.infer<typeof autonomousRunStateSchema>

export interface AutonomousLifecycleReasonView {
  code: string
  message: string
}

export interface AutonomousRunView {
  projectId: string
  agentSessionId: string
  runId: string
  runtimeKind: string
  runtimeLabel: string
  providerId: string
  supervisorKind: string
  status: AutonomousRunStatusDto
  statusLabel: string
  recoveryState: AutonomousRunRecoveryStateDto
  recoveryLabel: string
  duplicateStartDetected: boolean
  duplicateStartRunId: string | null
  duplicateStartReason: string | null
  startedAt: string
  lastHeartbeatAt: string | null
  lastCheckpointAt: string | null
  pausedAt: string | null
  cancelledAt: string | null
  completedAt: string | null
  crashedAt: string | null
  stoppedAt: string | null
  pauseReason: AutonomousLifecycleReasonView | null
  cancelReason: AutonomousLifecycleReasonView | null
  crashReason: AutonomousLifecycleReasonView | null
  lastErrorCode: string | null
  lastError: RuntimeRunDiagnosticDto | null
  updatedAt: string
  isActive: boolean
  isTerminal: boolean
  isStale: boolean
}

export interface AutonomousRunInspectionView {
  autonomousRun: AutonomousRunView | null
}

export function getAutonomousRunStatusLabel(status: AutonomousRunStatusDto): string {
  switch (status) {
    case 'starting':
      return 'Starting'
    case 'running':
      return 'Running'
    case 'paused':
      return 'Paused'
    case 'cancelling':
      return 'Cancelling'
    case 'cancelled':
      return 'Cancelled'
    case 'stale':
      return 'Stale'
    case 'failed':
      return 'Failed'
    case 'stopped':
      return 'Stopped'
    case 'crashed':
      return 'Crashed'
    case 'completed':
      return 'Completed'
  }
}

export function getAutonomousRunRecoveryLabel(recoveryState: AutonomousRunRecoveryStateDto): string {
  switch (recoveryState) {
    case 'healthy':
      return 'Healthy'
    case 'recovery_required':
      return 'Recovery required'
    case 'terminal':
      return 'Terminal'
    case 'failed':
      return 'Failed'
  }
}

function getAutonomousRunLabel(runtimeKind: string, status: AutonomousRunStatusDto): string {
  return `${normalizeText(runtimeKind, 'runtime')} · ${getAutonomousRunStatusLabel(status)}`
}

function mapLifecycleReason(reason: AutonomousLifecycleReasonDto | null | undefined): AutonomousLifecycleReasonView | null {
  if (!reason) return null
  return {
    code: reason.code,
    message: reason.message,
  }
}

export function mapAutonomousRun(autonomousRun: AutonomousRunDto): AutonomousRunView {
  const runtimeKind = normalizeText(autonomousRun.runtimeKind, 'runtime')
  const isTerminal = ['cancelled', 'failed', 'stopped', 'crashed', 'completed'].includes(autonomousRun.status)

  return {
    projectId: autonomousRun.projectId,
    agentSessionId: autonomousRun.agentSessionId,
    runId: normalizeText(autonomousRun.runId, 'autonomous-run-unavailable'),
    runtimeKind,
    runtimeLabel: getAutonomousRunLabel(runtimeKind, autonomousRun.status),
    providerId: normalizeText(autonomousRun.providerId, 'provider-unavailable'),
    supervisorKind: normalizeText(autonomousRun.supervisorKind, 'supervisor-unavailable'),
    status: autonomousRun.status,
    statusLabel: getAutonomousRunStatusLabel(autonomousRun.status),
    recoveryState: autonomousRun.recoveryState,
    recoveryLabel: getAutonomousRunRecoveryLabel(autonomousRun.recoveryState),
    duplicateStartDetected: autonomousRun.duplicateStartDetected,
    duplicateStartRunId: normalizeOptionalText(autonomousRun.duplicateStartRunId),
    duplicateStartReason: normalizeOptionalText(autonomousRun.duplicateStartReason),
    startedAt: autonomousRun.startedAt,
    lastHeartbeatAt: normalizeOptionalText(autonomousRun.lastHeartbeatAt),
    lastCheckpointAt: normalizeOptionalText(autonomousRun.lastCheckpointAt),
    pausedAt: normalizeOptionalText(autonomousRun.pausedAt),
    cancelledAt: normalizeOptionalText(autonomousRun.cancelledAt),
    completedAt: normalizeOptionalText(autonomousRun.completedAt),
    crashedAt: normalizeOptionalText(autonomousRun.crashedAt),
    stoppedAt: normalizeOptionalText(autonomousRun.stoppedAt),
    pauseReason: mapLifecycleReason(autonomousRun.pauseReason),
    cancelReason: mapLifecycleReason(autonomousRun.cancelReason),
    crashReason: mapLifecycleReason(autonomousRun.crashReason),
    lastErrorCode: normalizeOptionalText(autonomousRun.lastErrorCode),
    lastError: autonomousRun.lastError ?? null,
    updatedAt: autonomousRun.updatedAt,
    isActive: ['starting', 'running', 'paused', 'cancelling', 'stale'].includes(autonomousRun.status),
    isTerminal,
    isStale: autonomousRun.status === 'stale',
  }
}

export function mapAutonomousRunInspection(autonomousState: AutonomousRunStateDto): AutonomousRunInspectionView {
  return {
    autonomousRun: autonomousState.run ? mapAutonomousRun(autonomousState.run) : null,
  }
}
